File: nars_core_java/nars/language/Implication.java
/*
 * Implication.java
 *
 * Copyright (C) 2008  Pei Wang
 *
 * This file is part of Open-NARS.
 *
 * Open-NARS is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Open-NARS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Open-NARS.  If not, see <http://www.gnu.org/licenses/>.
 */
package nars.language;

import java.util.*;

import nars.io.Symbols;
import nars.storage.Memory;

/**
 * A Statement about an Inheritance relation.
 */
public class Implication extends Statement {

    /**
     * Constructor with partial values, called by make
     * @param arg The component list of the term
     */
    protected Implication(ArrayList<Term> arg) {
        super(arg);
    }

    /**
     * Constructor with full values, called by clone
     * @param n The name of the term
     * @param cs Component list
     * @param con Whether it is a constant term
     * @param i Syntactic complexity of the compound
     */
    protected Implication(String n, ArrayList<Term> cs, boolean con, short i) {
        super(n, cs, con, i);
    }

    /**
     * Clone an object
     * @return A new object
     */
    public Object clone() {
        return new Implication(name, (ArrayList<Term>) cloneList(components), isConstant(), complexity);
    }

    /**
     * Try to make a new compound from two components. Called by the inference rules.
     * @param subject The first component
     * @param predicate The second component
     * @param memory Reference to the memory
     * @return A compound generated or a term it reduced to
     */
    public static Implication make(Term subject, Term predicate, Memory memory) {
        if ((subject == null) || (predicate == null)) {
            return null;
        }
        if ((subject == null) || (predicate == null)) {
            return null;
        }
        if ((subject instanceof Implication) || (subject instanceof Equivalence) || (predicate instanceof Equivalence)) {
            return null;
        }
        if (invalidStatement(subject, predicate)) {
            return null;
        }
        String name = makeStatementName(subject, Symbols.IMPLICATION_RELATION, predicate);
        Term t = memory.nameToListedTerm(name);
        if (t != null) {
            return (Implication) t;
        }
        if (predicate instanceof Implication) {
            Term oldCondition = ((Implication) predicate).getSubject();
            if ((oldCondition instanceof Conjunction) && ((Conjunction) oldCondition).containComponent(subject)) {
                return null;
            }
            Term newCondition = Conjunction.make(subject, oldCondition, memory);
            return make(newCondition, ((Implication) predicate).getPredicate(), memory);
        } else {
            ArrayList<Term> argument = argumentsToList(subject, predicate);
            return new Implication(argument);
        }
    }

    /**
     * Get the operator of the term.
     * @return the operator of the term
     */
    public String operator() {
        return Symbols.IMPLICATION_RELATION;
    }
}

File: nars_scala/src/main/scala/nars/language/Implication.scala
package nars.logic.language

import java.util._
import nars.io.Symbols
import nars.storage.Memory
import Implication._
//remove if not needed
import scala.collection.JavaConversions._
import CompoundTerm._
import Statement._

object Implication {

  /**
   * Try to make a new compound from two components. Called by the inference rules.
   * @param subject The first component
   * @param predicate The second component
   * @param memory Reference to the memory
   * @return A compound generated or a term it reduced to
   */
  def make(subject: Term, predicate: Term, memory: Memory): Implication = {
    if ((subject == null) || (predicate == null)) {
      return null
    }
    if ((subject == null) || (predicate == null)) {
      return null
    }
    if ((subject.isInstanceOf[Implication]) || (subject.isInstanceOf[Equivalence]) || 
      (predicate.isInstanceOf[Equivalence])) {
      return null
    }
    if (invalidStatement(subject, predicate)) {
      return null
    }
    val name = makeStatementName(subject, Symbols.IMPLICATION_RELATION, predicate)
    val t = memory.nameToListedTerm(name)
    if (t != null) {
      return t.asInstanceOf[Implication]
    }
    if (predicate.isInstanceOf[Implication]) {
      val oldCondition = predicate.asInstanceOf[Implication].getSubject
      if ((oldCondition.isInstanceOf[Conjunction]) && 
        oldCondition.asInstanceOf[Conjunction].containComponent(subject)) {
        return null
      }
      val newCondition = Conjunction.make(subject, oldCondition, memory)
      make(newCondition, predicate.asInstanceOf[Implication].getPredicate, memory)
    } else {
      val argument = argumentsToList(subject, predicate)
      new Implication(argument)
    }
  }
}

/**
 * A Statement about an Inheritance relation.
 */
class Implication protected (arg: ArrayList[Term]) extends Statement(arg) {

  /**
   * Constructor with full values, called by clone
   * @param n The name of the term
   * @param cs Component list
   * @param con Whether it is a constant term
   * @param i Syntactic complexity of the compound
   */
  protected def this(n: String, 
      cs: ArrayList[Term], 
      con: Boolean, 
      i: Short) {
//    super(n, cs, con, i)
    this(cs)
    setName(n)
    this.isConstant_ = con
    this.complexity = i
  }

  /**
   * Clone an object
   * @return A new object
   */
  override def clone(): AnyRef = {
    new Implication(name, cloneList(components).asInstanceOf[ArrayList[Term]], isConstant_, complexity)
  }

  /**
   * Get the operator of the term.
   * @return the operator of the term
   */
  def operator(): String = Symbols.IMPLICATION_RELATION
}
